sailfishusermanager.cpp

```diff
@@ -0,0 +1,429 @@
+        return false;
+    }
+
+    struct lu_ent *ent_group = lu_ent_new();
+    if (lu_group_default(context, group.toLocal8Bit(), false, ent_group)) {
+        if (!lu_group_add(context, ent_group, &error)) {
+            qWarning() << "Adding group failed:" << lu_strerror(error);
+            rv = false;
+        }
+    } else {
+        qWarning() << "Getting group defaults failed";
+        rv = false;
+    }
+
+    lu_ent_free(ent_group);
+    lu_end(context);
+
+    return rv;
+}
+
+bool SailfishUserManager::addUserToGroup(const QString& user, const QString& group)
+{
+    struct lu_error *error = nullptr;
+    struct lu_context *context = lu_start(NULL, lu_user, NULL, NULL, lu_prompt_console, NULL, &error);
+    if (!context) {
+        qWarning() << "Error creating context";
+        return false;
+    }
+
+    bool rv = true;
+    struct lu_ent *ent = lu_ent_new();
+    if (lu_group_lookup_name(context, group.toLocal8Bit(), ent, &error)) {
+        GValue *value = g_slice_new0(GValue);
+        g_value_init(value, G_TYPE_STRING);
+        g_value_set_string(value, user.toLocal8Bit());
+        lu_ent_add(ent, LU_MEMBERNAME, value);
+        if (!lu_group_modify(context, ent, &error)) {
+            qWarning() << "Group modify failed:" << lu_strerror(error);
+            rv = false;
+        }
+    } else {
+        qWarning() << "Could not find group";
+        rv = false;
+    }
+
+    lu_ent_free(ent);
+    lu_end(context);
+
+    return rv;
+}
+
+bool SailfishUserManager::addUserToGroups(const QString& user)
+{
+    QFile file(GROUPS_IDS_FILE);
+    if (!file.open(QIODevice::ReadOnly)) {
+        qWarning() << "Failed to open groups file";
+        return false;
+    }
+
+    QByteArray line;
+
+    while (!file.atEnd()) {
+        line = file.readLine();
+        if (line.startsWith(GROUPS_USER)) {
+            foreach (const QString& group, line.mid(strlen(GROUPS_USER) + 1).split(',')) {
+                if (!addUserToGroup(user, group.trimmed())) {
+                    file.close();
+                    return false;
+                }
+            }
+        }
+    }
+    file.close();
+
+    // Droid groups
+    QFile droidDefault("/usr/lib/droid/hw-group.d/default.list");
+    if (droidDefault.open(QIODevice::ReadOnly)) {
+        while (!droidDefault.atEnd()) {
+            line = droidDefault.readLine();
+            if (!line.isEmpty()) {
+                if (!addUserToGroup(user, line.trimmed())) {
+                    droidDefault.close();
+                    return false;
+                }
+            }
+        }
+        droidDefault.close();
+    }
+    QFile droidInet("/usr/lib/droid/hw-group.d/inet.list");
+    if (droidInet.open(QIODevice::ReadOnly)) {
+        while (!droidInet.atEnd()) {
+            line = droidInet.readLine();
+            if (!line.isEmpty()) {
+                if (!addUserToGroup(user, line.trimmed())) {
+                    droidInet.close();
+                    return false;
```

Comment:
This return false on first failure logic -> a lot of things might have been despite the error. Should we care enough to attempt rollback? If not, would it be better/worse to do as much as possible before returning "it failed"?

Reply:
Let's add rollbacks later if testing finds them to be needed. Requires quite a lot of work for them to cover every possible situation.